modules/node-shared/src/main/scala/io/constellationnetwork/node/shared/http/routes/SnapshotRoutes.scala

```diff
@@ -33,124 +37,149 @@ final case class SnapshotRoutes[F[_]: Async, S <: Snapshot: Encoder, C: Encoder]
+      case _               => ServiceUnavailable(("message" ->> "Node is not ready yet") :: HNil)
+    }
+
+  /** Cached responses are already pre-serialized — no blocking work needed. */
+  private def respondWithCachedCombined(
+    opt: Option[LatestCombined],
+    byteSelector: LatestCombined => Array[Byte]
+  ): F[Response[F]] =
+    opt match {
+      case Some(cached) => Ok(Stream.emits(byteSelector(cached)).covary[F], jsonContentType)
+      case None         => NotFound()
+    }
+
+  private def respondWithCachedEntry(
+    opt: Option[CachedEntry],
+    byteSelector: CachedEntry => Array[Byte]
+  ): F[Response[F]] =
+    opt match {
+      case Some(cached) => Ok(Stream.emits(byteSelector(cached)).covary[F], jsonContentType)
+      case None         => NotFound()
+    }
+
+  private def respondWithStreamOption[A: Encoder](opt: Option[A]): F[Response[F]] =
+    opt match {
+      case Some(data) => Ok(streamJson(data), jsonContentType)
+      case None       => NotFound()
+    }
 
   protected val httpRoutes: HttpRoutes[F] = Timeout(snapshotTimeoutsConfig.routes)(HttpRoutes.of[F] {
+
     case GET -> Root / "latest" / "ordinal" =>
-      nodeStorage.getNodeState
-        .map(validStateForSnapshotReturn)
-        .ifM(
-          snapshotStorage.headSnapshot.map(_.map(_.ordinal)).flatMap {
-            case Some(ordinal) => Ok(("value" ->> ordinal.value.value) :: HNil)
-            case None          => NotFound()
-          },
-          serviceUnavailableNodeNotReady
-        )
+      whenNodeReady {
+        snapshotStorage.headSnapshot
+          .map(_.map(s => ("value" ->> s.ordinal.value.value) :: HNil))
+          .flatMap(respondWithStreamOption(_))
+      }
 
     case GET -> Root / "latest" / "metadata" =>
-      nodeStorage.getNodeState
-        .map(validStateForSnapshotReturn)
-        .ifM(
-          snapshotStorage.headSnapshot
-            .flatMap(_.traverse(snapshot => hasherSelector.forOrdinal(snapshot.ordinal)(implicit hasher => snapshot.toHashed[F])))
-            .map(_.map(snapshot => SnapshotMetadata(snapshot.ordinal, snapshot.hash, snapshot.lastSnapshotHash)))
-            .flatMap {
-              case Some(metadata) => Ok(metadata)
-              case None           => NotFound()
-            },
-          serviceUnavailableNodeNotReady
-        )
+      whenNodeReady {
+        latestSnapshotCache.getCombined
+          .map(_.map(_.metadata))
+          .flatMap(respondWithStreamOption(_))
+      }
 
     case req @ GET -> Root / "latest" =>
-      nodeStorage.getNodeState
-        .map(validStateForSnapshotReturn)
-        .ifM(
-          resolveEncoder[F, Signed[S]](req) { implicit enc =>
-            snapshotStorage.headSnapshot.flatMap {
-              case Some(snapshot) => Ok(snapshot)
-              case _              => NotFound()
-            }
-          },
-          serviceUnavailableNodeNotReady
-        )
+      whenNodeReady {
+        resolveEncoder[F, Signed[S]](req) { _ =>
```

Comment:
this `resolveEncoder` doesn't seem like it is necessary if there is only a single element in the `List[EntityEncoder[F, A]]` (jsonEncoders above) perhaps it is deprecated due to a new approach but I only see this `SnapshotRoutes` using it and like I said, if only one element then perhaps we could simplify the approach.

But maybe I am missing something here.